oci-image/src/state.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{Error, SnapshotType};
+use oci_spec::image::ImageConfiguration;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::{File, OpenOptions};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// The `MetadataManager` defines the contract that must be implemented by
+/// our state implementation.
+pub(crate) trait MetadataManager {
+    /// Add an `Image` to the state.
+    fn add_image(&mut self, image: &ImageMetadata) -> &mut Self;
+    /// Add a `Layer` to the state.
+    fn add_layer(&mut self, layer: &LayerMetadata) -> &mut Self;
+    /// Add a snapshot to the state.
+    fn add_snapshot(&mut self, snapshot_type: &SnapshotType, index: usize) -> &mut Self;
+    /// Check if the state contains an `Image` by it's identifier
+    fn has_image(&self, image_id: &str) -> bool;
+    /// Check if the state contains a `Layer` by it's identifier
+    fn has_layer(&self, layer_id: &str) -> bool;
+    /// Check if the state contains a snapshot
+    fn has_snapshot(&self, snapshot_id: &str) -> bool;
+    /// Get an image from the state by it's identifier
+    fn image(&self, image_id: &str) -> Option<&ImageMetadata>;
+    /// Get a layer from the state by it's identifier
+    fn layer(&self, layer_id: &str) -> Option<&LayerMetadata>;
+    // Get a snapshot index for the current snapshot_type
+    fn snapshot_index(&self, snapshot_type: &SnapshotType) -> usize;
+}
+
+/// `LayerMetadata` struct holds information's about a layer in the state.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct LayerMetadata {
+    /// The id of the layer
+    pub id: String,
+    /// The compressed layer digest
+    pub compressed_digest: String,
+    /// The decompressed layer digest,
+    pub uncompressed_digest: String,
+    /// The path where the layer is stored
+    pub store_path: String,
+}
+
+/// `ImageMetadata` struct holds information's about an image in the state.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct ImageMetadata {
+    /// The image id
+    pub id: String,
+    /// The image reference, such as `docker.io/amd64/ubuntu`
+    pub reference: String,
+    /// The image digest
+    pub digest: String,
+    /// The image layers metas
+    pub layers: Vec<LayerMetadata>,
+    /// The image configuration
+    pub config: ImageConfiguration,
+}
+
+/// `State` is responsible about storing information's about images and layers.
+#[derive(Clone, Default, Deserialize, Debug, Serialize, PartialEq)]
+pub(crate) struct State {
+    /// An hashmap that holds every images pulled
+    images: HashMap<String, ImageMetadata>,
+    /// An hashmap that holds every layers pulled
+    layers: HashMap<String, LayerMetadata>,
+    /// An hashmap that holds the current index for a specific snapshot type
```

Comment:
Why do you need to make this index specific to a snaphotter? Why not simply having a `sandbox_index: usize`that basically increments for every mount that is being done through the `ImageManager`, regardless of the snapshotter used?

`ImageManager`gets the next snapshot_index from its state, and passes it to the mount method of the snapshotter. Everytime state is being asked for a snapshot index, it gives the current one and increments it to fulfill the next request.